Disable password authentication when ssh keys have been set

markspanbroek opened the pull request:
Fixes #1058. The ssh daemon is started with the `PasswordAuthentication=no` option when an ssh key has been added to the config.json.

---
### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [x] Changes have been tested
- [x] `Change-type` present on at least one commit
- [x] `Signed-off-by` is present
- [x] The PR complies with the [Open Embedded Commit Patch Message Guidelines](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)
<!-- optional: `Changelog-entry` present on at least one commit if you want to set the changelog entry manually-->

### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

resin-jenkins commented:
Can one of the admins verify this patch?

ZubairLK commented:
@markspanbroek can you please elaborate your use-case/scenario a bit more?

This is a behavioural change that will impact the development images..

markspanbroek commented:
Hi @ZubairLK, my use case is basically the same as [this one](https://github.com/balena-os/meta-balena/issues/1058#issue-316852161); I'd like to secure my dev image so that others won't be able to login or push to my development board. That's why I implemented [this suggestion](https://github.com/balena-os/meta-balena/issues/1058#issuecomment-384788347). Does that answer your question?

ZubairLK commented:
@markspanbroek I see.

Some background: We are trying to work our way towards unifying prod/development images. So that we are able to switch these at runtime (or probably with a reboot). It requires some work though. 

- Being able to enable/disable getty services.
- Being able to change the kernel cmdline

Among other things.

e.g. We added support to ssh into production images via the users ssh key stored in their balenaCloud account. You can use a production image and just ssh to it locally using 
`ssh user@ip -p 22222` where `user` is your balenaCloud user.

As it stands, I don't think we can make this behavioural change in this way.

This needs a bit of thought. I'll ask internally. Instead of making it dependent on the existence of a custom ssh key. A better way would probably be to enable/disable Passwordless ssh via a flag in config.json. We will then be able to use that later on to switch between dev/prod images via the supervisor/dashboard.

markspanbroek commented:
Thanks for the explanation of what you're working towards, @ZubairLK. It would be a small effort on my side to write the code for enabling/disabling passwordless ssh via a new flag in config.json. Let me know if you think that that's a good idea, then I'll start with the implementation.

ZubairLK commented:
@markspanbroek 

Had an internal discussion. We'll take a step towards a flag in `config.json`.

e.g.
```
"os": {
     "sshWithoutPassword": true or false
}
```

This flag is supposed to only work in `development` images!

We don't want a user to be able to enable passwordless ssh in a production image and then 'forget'.

Defaults to `true` in `development` images.
And then on a case by case basis people can switch the flag to `false` in `development` images if needed.

I guess a fallback if not defined in `config.json` would be to check the etc/os-release for `VARIANT_ID` if its dev/prod.

This takes us one step closer to being able to switch dev/production images at runtime.

From a cursory look, it seems like sshd doesn't support config fragment files. We might have to use the service `-o option` to override the defaults in sshd_config_readonly with things read from `config.json`.

Thanks a lot for looking into this.

markspanbroek commented:
Hi @ZubairLK, I think I addressed all requirements with the latest commit.

> This flag is supposed to only work in `development` images!
> Defaults to `true` in `development` images.

If found that this was already addressed in [openssh_%.bbappend](https://github.com/balena-os/meta-balena/blob/92ece55590e4432e2145194d8bbcbfc99f22d767/meta-balena-common/recipes-connectivity/openssh/openssh_%25.bbappend#L45). I left that code as it is.

> And then on a case by case basis people can switch the flag to `false` in `development` images if needed.

The additional code from this pull request allows setting the flag to 'false'. It errs on the side of safety, so it can only disable password(less) authentication, not enable it.

> I guess a fallback if not defined in `config.json` would be to check the etc/os-release for `VARIANT_ID` if its dev/prod.

Turns out that I didn't need this, because the existing code already checked for the 'development-image' distro feature.

> From a cursory look, it seems like sshd doesn't support config fragment files. We might have to use the service `-o option` to override the defaults in sshd_config_readonly with things read from `config.json`.

I added an extra `-o` option using the `/home/root/.ssh/sshd.env` environment file.

Please let me know what you think.

alexgg commented:
hi @markspanbroek just a note to thank you again for this contribution and to assure you it's not forgotten and we'll review it shortly.

alexgg commented:
@balena-ci rebase

alexgg commented:
This is being addressed as part of https://github.com/balena-os/meta-balena/pull/2227

alexgg commented:
This has been addressed in https://github.com/balena-os/meta-balena/pull/2227 and will be released once device types update to meta-balena 2.85.0.

tl;dr development mode is now a runtime configuration. When configured in `config.json` it allows unauthenticated SSH logins, unless `os.sshKeys` are present in which case only authenticated SSH logins are allowed.
